### PR TITLE
Port Ryu float-to-string conversion algorithm to Scala

### LINKED ISSUE #11: Support Float / Double .toString
Basically it should follow the ECMAScript [6.1.6.1.20 Number::toString](https://tc39.es/ecma262/#sec-numeric-types-number-tostring) (where radix = 10).

We can simply to convert a floating-point number to a string like [Floating Point to Decimal Conversion is Easy](https://research.swtch.com/ftoa), and it is known that for roundtrip conversions (binary → decimal → binary), **17 decimal digits are sufficient** to recover the original value in IEEE 754 double-precision floating-point format [^1].

However, when converting a floating-point number to its decimal representation, it is widely accepted that algorithm should satisfy the following three properties [^dragon4] (it seems that ECMAScript spec also expect these(?))

- **No Information Loss**: The original value can be exactly reconstructed from the output string.
- **Shortest Output**: The string should be as short as possible without losing information.
- **Correct Rounding**: The conversion must follow IEEE 754 rounding rules.

There are several known algorithms to satisfy those traits.

- Dragon4 [^dragon4]
  - **Requires Multiple-Precision Integer (BigInteger)**
  - This isn't desirable for `Float.toString`: Converting floating point numbers to strings is a core language feature, whereas `BigInteger` is provided by `javalib`. It is undesirable for the language core to depend on the standard library.
- Grisu2/Grisu3 [^grisu]
  - The `Grisu` algorithm replaces the multi-precision integer arithmetic used in `Dragon4` with simulated high-precision floating-point arithmetic using integers. As a result, **it relies only on fixed-size integers (64-bit integers)**, making it faster than `Dragon4`.
  - `Grisu2` is an algorithm that produces the shortest decimal representation in most cases (according to the original paper, for approximately 99.5% of all IEEE 754 double-precision values). However, for a minor cases of inputs, it may return a non-shortest (but still correct and roundtrip-safe) decimal representation.
  - `Grisu3` is essentially the same as Grisu2 but with a stricter requirement: it fails when it cannot produce the shortest decimal representation. In such cases, the implementation is expected to fall back to another algorithm, such as `Dragon4`, to ensure it is shortest.
- Ryu [^ryu]
  - **It uses only fixed-size integers**, and even faster than `Grisu3`.
  - However, it relies on a relatively large lookup table [something like this](https://github.com/scala-native/scala-native/blob/2f7488ff0511746d370161853ec22101006c6b8f/javalib/src/main/scala/scala/scalanative/runtime/ieee754tostring/ryu/RyuDouble.scala#L62-L694).
  - [ScalaNative uses Ryu](https://github.com/scala-native/scala-native/tree/2f7488ff0511746d370161853ec22101006c6b8f/javalib/src/main/scala/scala/scalanative/runtime/ieee754tostring/ryu)

There're newer algorithms such as `Grisu-Exact`, `Schubfach`, and `DragonBox`.

If the "Shortest Output" is not always required, `Grisu2` would be a good option because of it's simplicity.
However, if we always want the shortest output, we cannot use `Grisu` and maybe `Ryu` is the option.

[^dragon4]: Guy L. Steele and Jon L. White. 1990. How to print floating-point numbers accurately. SIGPLAN Not. 25, 6 (Jun. 1990), 112–126. DOI:https://doi.org/10.1145/93548.93559

[^grisu]: Florian Loitsch. 2010. Printing floating-point numbers quickly and accurately with integers. SIGPLAN Not. 45, 6 (June 2010), 233–243. DOI:https://doi.org/10.1145/1809028.1806623

[^ryu]: Ulf Adams. 2018. Ryū: fast float-to-string conversion. SIGPLAN Not. 53, 4 (April 2018), 270–282. DOI:https://doi.org/10.1145/3296979.3192369

[^1]: David W. Matula. 1968. In-and-out conversions. *Commun. ACM* 11, 1 (Jan. 1968), 47–50. DOI: https://doi.org/10.1145/362851.362887

### PR DESCRIPTION
**Pull request by tanishiking**
close https://github.com/scala-wasm/scala-wasm/issues/11

Implement the Ryu algorithm for converting binary floating-point numbers the decimal string representation, as described in "Ryu: fast float-to-string conversion" by Ulf Adams (PLDI 2018). https://dl.acm.org/doi/10.1145/3192366.3192369

This port is based on the reference Java implementation from https://github.com/ulfjack/ryu with some comments and changes.

Most of the changes are addition of `RyuDouble` and `RyuFloat`, and Emitter to use those function for `toStringForConcat`. Rest of the change is adding a big pre-computed table, and test cases.

### CODE REVIEW

**Comment by tanishiking** [comment]
@sjrd Could you take a quick look especially on changes under `linker` and `test-suite`?
The core implementation is ported from https://github.com/ulfjack/ryu with some refactoring + comments on it.

**Comment by plokhotnyuk** [comment]
@tanishiking Have you considered to use more efficient algorithm from Raffaello Giulietti?

"The Schubfach way to render doubles":
https://drive.google.com/file/d/1luHhyQF9zKlM8yJ1nebU0OgVYhfC6CBN/view

Java sources with the license: 
https://github.com/c4f7fcce9cb06515/Schubfach/tree/3c92d3c9b1fead540616c918cdfef432bca53dfa/todec/src/math

Scala implementation that is based on it:
https://github.com/plokhotnyuk/jsoniter-scala/blob/c60560deb71a5f24b99d395f4b823569aa05a601/jsoniter-scala-core/jvm-native/src/main/scala/com/github/plokhotnyuk/jsoniter_scala/core/JsonWriter.scala#L2238-L2536

BTW, this algorithm is already implemented in the latest versions of JDK

**Comment by tanishiking** [comment]
Hi @plokhotnyuk,
I reviewed papers for Dragon4, Grisu3, and Ryu, and chose Ryu for its completeness, and since it's used in `scala-native` as well. I didn't seriously consider more recent algorithms like Schubfach and Dragonbox. 😞 
However, after looking at `jsoniter-scala`, the Java implementation, and take a quick look at the paper, Schubfach seems to be a very elegant algorithm and would definitely be worth adopting (whereas Dragonbox seems somewhat more complicated).
And indeed, Schubfach seems to be faster than Ryu https://github.com/abolz/Drachennest

That said, performance isn't our top priority at the moment, so I feel like continuing with tried and tested Ryu for now and consider migrating to Schubfach in the future (perhaps in scala-native as well 🙌 ).

Anyway, thanks for sharing @plokhotnyuk !

File: linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala [hunk 1]
```diff
@@ -2322,16 +2322,18 @@ private class FunctionEmitter private (
             }
           case FloatType =>
             if (targetPureWasm) {
-              fb += wa.Drop
-              fb ++= ctx.stringPool.getConstantStringInstr("0")
+              fb += wa.Call(genFunctionID.forMethod(MemberNamespace.PublicStatic,
+                  SpecialNames.RyuFloatClass, SpecialNames.floatToStringMethodName))
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/FunctionEmitter.scala]: By spec, we must use `doubleToString` even for floats. Otherwise `x.toString()` won't give the same result as `(x: Any).toString()`.

So we really don't need the variant for `Float`s. We only use the one for `Double`s.

File: linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/Emitter.scala [hunk 2]
```diff
@@ -587,7 +587,11 @@ object Emitter {
 
       // Implementation of Float_% and Double_%
       callStaticMethod(WasmRuntimeClass, fmodfMethodName),
-      callStaticMethod(WasmRuntimeClass, fmoddMethodName)
+      callStaticMethod(WasmRuntimeClass, fmoddMethodName),
+
+      // TODO Ideally we should not require these, but rather adapt to their absence
+      callStaticMethod(RyuFloatClass, floatToStringMethodName),
+      callStaticMethod(RyuDoubleClass, doubleToStringMethodName)
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/backend/wasmemitter/Emitter.scala]: These should be in the `linker-private-library`, like `fmod`. And those requirements (in fact just the `double` one, see my other comment) should be conditioned on `targetPureWebAssembly`. You can use `cond` for that, like we do for error classes 20 lines above.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/DoubleTest.scala [hunk 3]
```diff
@@ -63,32 +63,271 @@ class DoubleTest {
...
+  @inline
+  private def assertDoubleToString(expected: String, value: Double): Unit = {
+    /* The optimizer constant fold the string conversion of `<float literal>.toString()` directly within the JDK.
+     * However, because the rounding behavior of Double.toString differs between versions before and after JDK 19.
+     * For instance, in JDK 21, `Double.toString(double)` specifies "round to even" for boundary cases,
+     * whereas this specific boundary behavior seems not specified in JDK 17.
+     * https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/lang/Double.html#toString(double)
+     * https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/lang/Double.html#toString(double)
+     * To ensure stability, we are skipping comparisons against `value.toString()`, to avoid being depend on the
+     * specific JDK's string conversion implementation.
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/DoubleTest.scala]: This comment, in fact, explains why it's invalid to constant-fold `floatLit.toString()` on the JDK. We can't do that.

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala [hunk 4]
```diff
@@ -4212,17 +4212,30 @@ private[optimizer] abstract class OptimizerCore(
       def constant(s: String): PreTransform =
         PreTransLit(StringLiteral(s))
 
-      def forFloatingPoint(value: Double): PreTransform =
-        jsNumberToString(value).fold(preTrans)(s => constant(s))
+      def forFloat(value: Float): PreTransform = {
+        if (targetPureWasm) {
+          constant(value.toString)
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala]: As the ignored test demonstrates, this is an invalid optimization. We must never constant-fold using the JDK's `floatValue.toString()`.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/DoubleTest.scala [hunk 5]
```diff
@@ -63,32 +63,271 @@ class DoubleTest {
...
     }
-    assertEquals("1.2", 1.2.toString)
+    assertDoubleToString("1.2", 1.2)
   }
 
-  @Test def toHexStringTest(): Unit = {
-    assumeFalse("Floating point to string doesn't work in pure Wasm.",
+  @Test def toStringTest(): Unit = {
+    assumeTrue(
+        "The output of JS and JVM (depends on JDK version) differs slightly from the Ryu implementations.",
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/DoubleTest.scala]: Could we tweak Ryu's boundary conditions to always do the same as JS? JS has been well-specified since forever, and that's already what Scala.js-on-Wasm uses.

File: javalib/src/main/scala/java/lang/flttodec/RyuDoubleTables.scala [hunk 6]
```diff
@@ -0,0 +1,720 @@
...
+  val PosTableSize = 326
+  val NegTableSize = 291
+  val Pow5BitCount = 121
+  val Pow5QuarterBitCount = 31
+  val Pow5InvBitCount = 122
+  val Pow5InvQuarterBitCount = 31
+
+  /** Pow5Split contains pre-computed 5^q values split into 4 31-bit parts. */
+  val Pow5Split = Array(
+    ( 134217728,          0,          0,          0),
```
Comment (sjrd) [inline javalib/src/main/scala/java/lang/flttodec/RyuDoubleTables.scala]: This boxes all the ints. We don't want that here.

We could define a `class FourInts` with 4 `Int` fields (anyway you won't have `Tuple`s available in the linker-private-library). But even that will create a lot of unnecessary code in the generated Wasm.

Consider using a single, flat `Array[Int]`, and index into with `index * 4 + part`.

File: javalib/src/main/scala/java/lang/flttodec/RyuDouble.scala [hunk 7]
```diff
@@ -0,0 +1,446 @@
...
+        removed += 1
+      }
+      bq + (if ((bq == aq) || (lastRemovedDigit >= 5)) 1 else 0)
+    }
+    (output, removed)
+  }
+
+  @inline
+  private def printString(output0: Long, olength: Int, exp0: Int,
+      sign: Boolean, scientificNotation: Boolean): String = {
```
Comment (sjrd) [inline javalib/src/main/scala/java/lang/flttodec/RyuDouble.scala]:
```suggestion
  private def printString(output0: Long, olength: Int, exp0: Int,
      sign: Boolean, scientificNotation: Boolean): String = {
```
This large method does not return a tuple, so I guess it's counter-productive to inline it.

File: javalib/src/main/scala/java/lang/flttodec/RyuDouble.scala [hunk 8]
```diff
@@ -0,0 +1,446 @@
...
+    var count = 4
+    var v = value / 625
+    while (v > 0) {
+      if (v % 5 != 0) {
+        return count
+      }
+      v /= 5
+      count += 1
+    }
+    throw new IllegalArgumentException(s"$value")
```
Comment (sjrd) [inline javalib/src/main/scala/java/lang/flttodec/RyuDouble.scala]: `AssertionError`? Same for the other ones. If we run into an exception anywhere in this conversion, it's a bug in the conversion, not a user-space mistake.

**Review by sjrd — CHANGES_REQUESTED** [review]

File: javalib/src/main/scala/java/lang/flttodec/RyuDouble.scala (hunk 7 above)
Comment (tanishiking) [reply javalib/src/main/scala/java/lang/flttodec/RyuDouble.scala]: Since `printString` is called only from `doubleToString`, I guess it's not counter-productive I guess?
(The original Ryu code has inlined into the `doubleToString`, but I split into a function for readability) (embedding the printString into `doubleToString` is the best for performance maybe though)

**Comment by tanishiking** [comment]
Thanks @sjrd for review!
Could you take another look if all feedbacks are fixed?

- Use `doubleToString` even for `Float`
- Don't constant fold in JVM
- Move RyuDouble to `linker-private-library` and requires it only when targeting pure Wasm.
- Use flat `Array[Int]`
- Align the output of Ryu with JS.
  - mostly about scientific notation behavior, and trim trailing `.0`.
- **Made ToStringTest to test roundtrip as well**.

Maybe you don't need to take a look RyuDouble implementation very carefully, since it's basically a port from reference implementation, and maybe we swap the implementation to another algorithm in future (?)

File: javalib/src/main/scala/java/util/internal/Tuples.scala [hunk 9]
```diff
@@ -21,6 +21,10 @@ final class Tuple3[+T1, +T2, +T3](val _1: T1, val _2: T2, val _3: T3)
 @inline
 final class Tuple4[+T1, +T2, +T3, +T4](val _1: T1, val _2: T2, val _3: T3, val _4: T4)
 
+@inline
+final class Tuple6[+T1, +T2, +T3, +T4, +T5, +T6](
+    val _1: T1, val _2: T2, val _3: T3, val _4: T4, val _5: T5, val _6: T6)
```
Comment (sjrd) [inline javalib/src/main/scala/java/util/internal/Tuples.scala]: This is now unused.

File: linker/jvm/src/main/scala/org/scalajs/linker/backend/emitter/PrivateLibHolder.scala [hunk 10]
```diff
@@ -31,6 +31,8 @@ object PrivateLibHolder {
       "org/scalajs/linker/runtime/UndefinedBehaviorError.sjsir",
       "org/scalajs/linker/runtime/WasmRuntime.sjsir",
       "org/scalajs/linker/runtime/WasmRuntime$.sjsir",
+      "org/scalajs/linker/runtime/RyuDouble.sjsir",
+      "org/scalajs/linker/runtime/RyuDouble$.sjsir",
```
Comment (sjrd) [inline linker/jvm/src/main/scala/org/scalajs/linker/backend/emitter/PrivateLibHolder.scala]: Put those between `RuntimeLong` and `UndefinedBehaviorError` to preserve the alphabetical order.

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala [hunk 11]
```diff
@@ -4221,8 +4221,8 @@ private[optimizer] abstract class OptimizerCore(
         case ShortLiteral(value)   => constant(value.toString)
         case IntLiteral(value)     => constant(value.toString)
         case LongLiteral(value)    => constant(value.toString)
-        case FloatLiteral(value)   => forFloatingPoint(value)
-        case DoubleLiteral(value)  => forFloatingPoint(value)
+        case FloatLiteral(value)  if !targetPureWasm => forFloatingPoint(value)
+        case DoubleLiteral(value) if !targetPureWasm => forFloatingPoint(value)
```
Comment (sjrd) [inline linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala]: This change is probably (hopefully) not necessary anymore, is it?

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/DoubleTest.scala [hunk 12]
```diff
@@ -64,31 +64,270 @@ class DoubleTest {
   }
 
   @Test def toStringWithIntegerValuesWhenAnInteger(): Unit = {
-    assumeFalse("Floating point to string doesn't work in pure Wasm.",
-        executingInPureWebAssembly)
     if (executingInJVM) {
       assertEquals("0.0", 0.0.toString)
-      assertEquals("-0.0", (-0.0).toString)
+      assertEquals("-0.0", -0.0.toString)
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/DoubleTest.scala]: Revert these changes to preserve the parens?

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/DoubleTest.scala [hunk 13]
```diff
@@ -64,31 +64,270 @@ class DoubleTest {
...
 
+  @Test def toStringTest(): Unit = {
+    assumeFalse("The tests are designed for JS behavior.", executingInJVM)
+
+    def test(expectedStr: String, value: Double): Unit = {
+      val actualStr = value.toString
+      assertEquals(expectedStr, actualStr)
+      // Test roundtrip: parsing the string should give back the exact same value
+      val parsed = JDouble.parseDouble(actualStr)
+      assertEquals(value, parsed, 0.0)
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/DoubleTest.scala]: Use `org.scalajs.testsuite.utils.AssertExtensions.assertExactEquals` here.

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/DoubleTest.scala [hunk 14]
```diff
@@ -64,31 +64,270 @@ class DoubleTest {
...
+
+    // switchToSubnormal
+    test("2.2250738585072014e-308", JDouble.longBitsToDouble(0x0010000000000000L))
+    /* Floating point values in the range 1.0e-7 < x < 1.0e+21 have to be printed
+     * without exponent. This test checks the values at those boundaries.
+     */
+
+    // boundaryConditions
+    // x = 1.0E7
+    test("10000000", 1.0E7d)
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/DoubleTest.scala]: Nit: be consistent with the case of the `e`s. In this codebase we prefer the lowercase version (notably because that's what `toString()` does).

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/DoubleTest.scala [hunk 15]
```diff
@@ -64,31 +64,270 @@ class DoubleTest {
...
+    test("100000000000000000000", 99999999999999999999.999999998d) // x < 1.0E20
+    test("0.000001", 0.000001d) // x = 1.0e-6
+    test("0.000009999999999999997", 0.000009999999999999997d) // x > 1.0e-7
+    test("1e-7", 0.0000001d) // x = 1.0e-7
+
+    // minAndMax
+    test("1.7976931348623157e+308", JDouble.longBitsToDouble(0x7fefffffffffffffL))
+    // TODO: (align the output between Ryu and JS)
+    assertEquals(
+      if (executingInPureWebAssembly) "4.9e-324" else "5e-324",
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/DoubleTest.scala]: Can we do that now, before we forget?

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/DoubleTest.scala [hunk 16]
```diff
@@ -64,31 +64,270 @@ class DoubleTest {
...
+    test("4.294967298", 4.294967298) // 2^32 + 2
+
+    // MinMaxShift
+    // Test min, max shift values in shiftright128
+    val maxMantissa = (1L << 53) - 1
+    // 32-bit opt-size=0:  49 <= dist <= 50
+    // 32-bit opt-size=1:  30 <= dist <= 50
+    // 64-bit opt-size=0:  50 <= dist <= 50
+    // 64-bit opt-size=1:  30 <= dist <= 50
+    assertEquals("1.7800590868057611e-307", ieeeParts2Double(sign = false, 4, 0).toString)
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/DoubleTest.scala]:
```suggestion
    test("1.7800590868057611e-307", ieeeParts2Double(sign = false, 4, 0))
```
and everywhere below?

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/FloatTest.scala [hunk 17]
```diff
@@ -73,25 +73,23 @@ class FloatTest {
     }
   }
 
-  @Test def toStringWithIntegerValuesWhenAnInteger(): Unit = {
-    assumeFalse("Floating point to string doesn't work in pure Wasm.",
-        executingInPureWebAssembly)
+@Test def toStringWithIntegerValuesWhenAnInteger(): Unit = {
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/FloatTest.scala]:
```suggestion
  @Test def toStringWithIntegerValuesWhenAnInteger(): Unit = {
```

File: test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/FloatTest.scala [hunk 18]
```diff
@@ -73,25 +73,23 @@ class FloatTest {
...
     if (executingInJVM) {
       assertEquals("0.0", 0.0f.toString)
-      assertEquals("-0.0", (-0.0f).toString)
+      assertEquals("-0.0", -0.0f.toString)
```
Comment (sjrd) [inline test-suite/shared/src/test/scala/org/scalajs/testsuite/javalib/lang/FloatTest.scala]: Same here as above: revert those changes?

File: linker-private-library/src/main/scala/org/scalajs/linker/runtime/RyuDouble.scala [hunk 19]
```diff
@@ -0,0 +1,1162 @@
...
+    val olength = vplength - removed
+
+    // Step 5: Print the decimal representation.
+    // We follow Double.toString semantics here.
+    printString(output, olength, exp, sign, scientificNotation)
+  }
+
+  @inline
+  private def convertBinaryToDecimalInterval(u: Long, v: Long, w: Long, e2: Int,
+      isGeneralLowerBoundCase: Boolean, even: Boolean): (Long, Long, Long, Int, Boolean, Boolean) = {
```
Comment (sjrd) [inline linker-private-library/src/main/scala/org/scalajs/linker/runtime/RyuDouble.scala]: Ouch, this shouldn't work. I think the `JavalibIRCleaner` is too lenient in the linker-private-library. It shouldn't allow references to our custom tuple classes from the javalib. The linker-private-library should always be able to link with any ABI-compliant version of the javalib.

Use a custom `@inline private final class` instead of the 6-tuple.

File: linker-private-library/src/main/scala/org/scalajs/linker/runtime/RyuDouble.scala [hunk 20]
```diff
@@ -0,0 +1,1162 @@
...
+      } else {
+        (false, false, cq)
+      }
+      (aq, bq, cq, q + e2, za, zb)
+    }
+  }
+
+  @inline
+  private def computeShortest(aq0: Long, bq0: Long, cq0: Long, za0: Boolean, zb0: Boolean,
+      e10: Int, even: Boolean, scientificNotation: Boolean): (Long, Int) = {
```
Comment (sjrd) [inline linker-private-library/src/main/scala/org/scalajs/linker/runtime/RyuDouble.scala]: Same here. Use a custom `@inline private final class` for the tuple.

**Review by sjrd — CHANGES_REQUESTED** [review]
Mostly nits.

There's a problem with the use of tuples in the linker-private-library.

There's also the case of `4.9` vs `5`. If it's not too much trouble, we should fix it now before we forget. If it's hard, let's open an issue to keep track of it.

File: linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala (hunk 11 above)
Comment (tanishiking) [reply linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/OptimizerCore.scala]: Yeah, now the RyuDouble basically prints the same as JavaScript, we can remove this 👍

**Comment by tanishiking** [comment]
> There's also the case of 4.9 vs 5. If it's not too much trouble, we should fix it now before we forget. If it's hard, let's open an issue to keep track of it.

Aligned the output with the JS in https://github.com/scala-wasm/scala-wasm/pull/116/commits/0ca3688debbacdaca9198fc4c89d034ef535ca99
It was just `RyuDouble#computeShortest` had a logic to keep at least two digits in scientific notation output because of JDK `Double.toString` does. Just removed that (because JS always print the shortest digits) :+1:

**Review by sjrd — APPROVED** [review]
Nice!

**Comment by plokhotnyuk** [comment]
@tanishiking [Today I adopted the XJB algorithm](https://github.com/plokhotnyuk/jsoniter-scala/commit/f7f48455ebf2b026e4511b3a6c5fcb816d538ceb). In practice this noticeably reduced CPU usage and sped up float/double serialization comparing to previous Schubfach-base implementation (up to 25% on JVM and up to 45% in JS browsers on real-world message samples).

**Comment by tanishiking** [comment]
@plokhotnyuk 
Wow, it's an algorithm that was just announced in a preprint recently? https://github.com/xjb714/f2dec_bench

I haven't checked the algorithm yet, but if it handles float-to-string conversion correctly, that's incredible 😮 (It's even faster than DragonBox) (and 25,45% speed up in JVM and JS is awesome!).

I'm curious that, from what I read in https://internals.rust-lang.org/t/algorithm-new-float-double-to-string-algorithm/23670/, it doesn't seem like they have verified the algorithms properties Information Preservation (round-tripping from decimal back to binary and vice versa), Minimum length, and Correct Rounding. Did you find any problem, or is it working good so far?

**Comment by plokhotnyuk** [comment]
> Did you find any problem, or is it working good so far?
I have used brute force and checked all `floats` - the output is always parsed without any bit of loss and representation is always p 



> @plokhotnyuk Wow, it's an algorithm that was just announced in a preprint recently? https://github.com/xjb714/f2dec_bench
>

Yes, I only skimmed through the preprint and switched to adopt and test it first ;)
 
> I haven't checked the algorithm yet, but if it handles float-to-string conversion correctly, that's incredible 😮 (It's even faster than DragonBox) (and 25,45% speed up in JVM and JS is awesome!).
> 
> I'm curious that, from what I read in https://internals.rust-lang.org/t/algorithm-new-float-double-to-string-algorithm/23670/, it doesn't seem like they have verified the algorithms properties Information Preservation (round-tripping from decimal back to binary and vice versa), Minimum length, and Correct Rounding. Did you find any problem, or is it working good so far?

I exhaustively tested all 2^32 possible float values: every one round-trips perfectly with no bit loss, and the textual representation is always the same length or shorter than Java’s Schubfach implementation in the latest JDKs. I also spent many CPU-hours stress-testing double values, and so far everything looks rock solid.

**Comment by tanishiking** [comment]
Sounds very promising!
I'll keep an eye on the project, and we might want to move from ryu to XJB in the future :) Thank you so much for sharing 🙏

**Comment by tanishiking** [comment]
Sounds very promising!
I'll keep an eye on the project, and we might want to move from ryu to XJB in the future :) Thank you so much for sharing 🙏
